edge-util/src/Microsoft.Azure.Devices.Edge.Storage.RocksDb/disk/DiskSpaceChecker.cs

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Storage.RocksDb.Disk
+{
+    using System;
+    using System.IO;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Storage.Disk;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+
+    public class DiskSpaceChecker : IStorageSpaceChecker
+    {
+        static readonly int checkFrequencyDefault = 3;
+        readonly string storageFolder;
+        readonly object updateLock = new object();
+        Option<DiskSpaceCheckerBase> inner;
+        long maxSize;
+        Option<int> checkFrequency;
+
+        DiskSpaceChecker(string storageFolder, Option<DiskSpaceCheckerBase> inner)
+        {
+            this.storageFolder = Preconditions.CheckNonWhiteSpace(storageFolder, nameof(storageFolder));
+            this.inner = Preconditions.CheckNotNull(inner, nameof(inner));
+        }
+
+        public bool IsFull => this.inner.Match<bool>(b => b.DiskStatus == DiskSpaceStatus.Full, () => false);
+
+        public static DiskSpaceChecker Create(string storageFolder)
+        {
+            // Start up diskSpaceChecker unfilled - so that it will be enabled when SetMaxSizeBytes is called
+            var diskSpaceChecker = new DiskSpaceChecker(storageFolder, Option.None<DiskSpaceCheckerBase>());
+            Events.Created(storageFolder);
+            return diskSpaceChecker;
+        }
+
+        public void SetMaxSizeBytes(long maxSizeBytes)
+        {
+            this.maxSize = maxSizeBytes;
+            this.EnableChecker();
+            Events.SetMaxSizeDiskSpaceUsage(this.maxSize, this.storageFolder);
+        }
+
+        public void SetCheckFrequency(Option<int> checkFrequencySecs)
+        {
+            checkFrequencySecs.ForEach(updatedCheckFrequency =>
+            {
+                if (updatedCheckFrequency > 0 && !this.checkFrequency.Equals(Option.Some(updatedCheckFrequency)))
+                {
+                    lock (this.updateLock)
+                    {
+                        this.checkFrequency = Option.Some(updatedCheckFrequency);
+                        this.inner.ForEach(b => b.DisposeChecker());
+                        this.inner = Option.Some(new FixedSizeDiskSpaceChecker(this.storageFolder, this.maxSize, TimeSpan.FromSeconds(updatedCheckFrequency), Events.Log) as DiskSpaceCheckerBase);
+                    }
+                }
+            });
+        }
+
+        void EnableChecker()
+        {
+            lock (this.updateLock)
+            {
+                this.inner = Option.Some(
+                    new FixedSizeDiskSpaceChecker(
+                        this.storageFolder,
+                        this.maxSize,
+                        TimeSpan.FromSeconds(this.checkFrequency.GetOrElse(checkFrequencyDefault)),
+                        Events.Log)
+                    as DiskSpaceCheckerBase);
```

Comment:
You should add a check before this to make sure we don't try to enable/initialize the checker again if it's already created.

if(!this.inner.HasValue)

Reply:
Added the check - which also trickled into changing a few other things.